Importando as bibliotecas e módulos necessários: 

In [4]:
import os
import pandas as pd
import numpy as np
import sys
import gc

!pip install openpyxl

Os datasets analisados neste notebook são as Bases de dados de Alunos: Os microdados de alunos do SAEB são apresentados em base de dados agrupadas segundo a série avaliada. Dessa forma, as bases de dados dos alunos são apresentadas separadamente, incluindo dados de respostas e resultados dos testes de Língua Portuguesa, de Matemática, de Ciências Humanas e de Ciências da Natureza, bem como as respostas aos questionários. Cada uma das bases traz o conjunto completo de alunos participantes naquela série. Analisaremos apenas os dados dos alunos do 9º ano do EF.

### 1. Coleta de dados 
Esta seção descreve o processo de coleta de dados realizado para construir os conjuntos de dados.

### 1.1 Estrutura do conjunto de dados
Esta seção descreve as estruturas e conteúdos dos conjuntos de dados utilizados, incluindo o número de linhas e colunas e descrições das observações
e variáveis. Eles foram carregados em um pandas.DataFrame, e como já estão fornecidos como um arquivo CSV, podem ser carregados facilmente usando a 
função read_csv().

#### SAEB 2019
O dataset de 2019 será tomado como base para os demais anos desse estudo.

In [5]:
gc.collect
df_saeb_2019 = pd.read_csv("2019/TS_ALUNO_9EF.csv",  encoding='latin-1', sep=";", dtype='category')
df_saeb_2019

ID_SAEB ID_REGIAO ID_UF ID_MUNICIPIO ID_AREA ID_ESCOLA IN_PUBLICA  \
0          2019         1    11      6311030       2  61255632          1   
1          2019         1    11      6311030       2  61255632          1   
2          2019         1    11      6311030       2  61255632          1   
3          2019         1    11      6311030       2  61255632          1   
4          2019         1    11      6311030       2  61255632          1   
...         ...       ...   ...          ...     ...       ...        ...   
2388926    2019         5    53      6316599       1  61322272          1   
2388927    2019         5    53      6316599       1  61322272          1   
2388928    2019         5    53      6316599       1  61322272          1   
2388929    2019         5    53      6316599       1  61322272          1   
2388930    2019         5    53      6316599       1  61322272          1   

        ID_LOCALIZACAO ID_TURMA ID_SERIE  ... TX_RESP_Q016 TX_RESP_Q017A  \
0                    1  1030927        9  ...            A             D   
1                    1  1030927        9  ...            A             D   
2                    1  1030927        9  ...            A             D   
3                    1  1030927        9  ...            A             C   
4                    1  1030927        9  ...            A             C   
...                ...      ...      ...  ...          ...           ...   
2388926              1  1067040        9  ...            .             .   
2388927              1  1067040        9  ...            A             D   
2388928              1  1067040        9  ...            A             D   
2388929              1  1067040        9  ...            A             D   
2388930              1  1067040        9  ...            .             .   

        TX_RESP_Q017B TX_RESP_Q017C TX_RESP_Q017D TX_RESP_Q017E TX_RESP_Q018A  \
0                   A             D             B             A             C   
1                   A             D             B             A             A   
2                   A             D             C             A             C   
3                   A             D             C             A             A   
4                   A             *             B             A             B   
...               ...           ...           ...           ...           ...   
2388926             .             .             .             .             .   
2388927             B             C             D             A             B   
2388928             A             A             C             A             B   
2388929             A             B             B             A             B   
2388930             .             .             .             .             .   

        TX_RESP_Q018B TX_RESP_Q018C TX_RESP_Q019  
0                   B             A            C  
1                   A             A            C  
2                   B             B            D  
3                   A             A            D  
4                   A             A            C  
...               ...           ...          ...  
2388926             .             .            .  
2388927             C             B            A  
2388928             C             B            C  
2388929             .             .            B  
2388930             .             .            .  

[2388931 rows x 129 columns]

Informações gerais sobre o conjunto de dados:
Como ele contém mais de 100 colunas (o valor padrão de display.max_info_columns), o parâmetro verbose de info() foi definido como True para exibir todas as colunas de dados. Pelo mesmo motivo, o parâmetro null_counts também foi explicitamente definido como True.

In [6]:
df_saeb_2019.info(verbose=True, show_counts = True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2388931 entries, 0 to 2388930
Data columns (total 129 columns):
 #    Column                         Non-Null Count    Dtype   
---   ------                         --------------    -----   
 0    ID_SAEB                        2388931 non-null  category
 1    ID_REGIAO                      2388931 non-null  category
 2    ID_UF                          2388931 non-null  category
 3    ID_MUNICIPIO                   2388931 non-null  category
 4    ID_AREA                        2388931 non-null  category
 5    ID_ESCOLA                      2388931 non-null  category
 6    IN_PUBLICA                     2388931 non-null  category
 7    ID_LOCALIZACAO                 2388931 non-null  category
 8    ID_TURMA                       2388931 non-null  category
 9    ID_SERIE                       2388931 non-null  category
 10   ID_ALUNO                       2388931 non-null  category
 11   IN_SITUACAO_CENSO              2388931 non-null 

O conjunto de dados contém 2388931 observações (linhas) em 129 variáveis (colunas). Cada observação corresponde a um aluno do 9° ano participante na 
avaliação em 2019, e as variáveis correspondem a coleta de informações que possam avaliar os eixos do SAEB como equidade (nível socioeconômico dos estudantes, perfil individual e familiar) e atendimento escolar (acesso e trajetória escolar) e desempenho nas provas.
As variáveis e esquemas de codificação para respostas categóricas selecionadas são os seguintes:

In [7]:
dicionario_2019 = pd.read_excel(r"2019/Dicionario_Saeb_2019.xlsx", sheet_name="TS_ALUNO_9EF")
dicionario_2019 = dicionario_2019[["TS_ALUNO_9EF", "Unnamed: 3"]]
dicionario_2019 = dicionario_2019.dropna(how="all")

from IPython.display import display
display(dicionario_2019.style.hide(axis="index"))

Abaixo está uma listagem literal das perguntas do Questionário de Estudante(TX_RESP_Q001 a TX_RESP_Q019). Os esquemas de codificação para as respostas categóricas também estão enumerados:
#### ID_SAEB
- 2019
#### ID_REGIAO	
- 1: Norte
- 2: Nordeste
- 3: Sudeste
- 4: Sul
- 5: Centro-Oeste
#### ID_UF	
- 11: RO
- 12: AC
- 13: AM
- 14: RR
- 15: PA
- 16: AP
- 17: TO
- 21: MA
- 22: PI
- 23: CE
- 24: RN
- 25: PB
- 26: PE
- 27: AL
- 28: SE
- 29: BA
- 31: MG
- 32: ES
- 33: RJ
- 35: SP
- 41: PR
- 42: SC 
- 43: RS
- 50: MS
- 51: MT
- 52: GO
- 53: DF
#### ID_AREA	
- 1: Capital
- 2: Interior
#### IN_PUBLICA	
- 0: Não pública
- 1: Pública
#### ID_LOCALIZACAO	
- 1: Urbana
- 2: Rural
#### ID_SERIE	
- 9: 9° ano do Ensino Fundamental
#### IN_SITUACAO_CENSO
- 0: Não consistente
- 1: Consistente
#### IN_PREENCHIMENTO_LP	
- 0: Prova não prenchida
- 1: Prova preenchida
#### IN_PREENCHIMENTO_MT
- 0: Prova não prenchida
- 1: Prova preenchida
#### IN_PREENCHIMENTO_CH	
- 0: Prova não prenchida
- 1: Prova preenchida
- Vazio: Não selecionado para a amostra de CH
#### IN_PREENCHIMENTO_CN	
- 0: Prova não prenchida
- 1: Prova preenchida
- Vazio: Não selecionado para a amostra de CN
#### IN_PRESENCA_LP	
- 0: Ausente
- 1: Presente
#### IN_PRESENCA_MT
- 0: Ausente
- 1: Presente
#### IN_PRESENCA_CH
- 0: Ausente
- 1: Presente
#### IN_PRESENCA_CN
- 0: Ausente
- 1: Presente
#### ID_CADERNO_LP
- Prova Regular (Cadernos 1 a 21)
- Macrotipo 18 (Caderno 22)
#### ID_BLOCO_1_LP	
- De 1 a 7 (8 no macrotipo 18 - caderno 22)
#### ID_BLOCO_2_LP	
- De 1 a 7 (9 no macrotipo 18 - caderno 22)
#### ID_CADERNO_MT
- Prova Regular (Cadernos 1 a 21)
- Macrotipo 18 (Caderno 22)
#### ID_BLOCO_1_MT	
- De 1 a 7 (8 no macrotipo 18 - caderno 22)
#### ID_BLOCO_2_MT
- De 1 a 7 (9 no macrotipo 18 - caderno 22)
#### ID_CADERNO_CH
- Prova Regular (Cadernos 1 a 26)
- Macrotipo 18 (Caderno 27)
#### ID_BLOCO_1_CH	
- De 1 a 13
#### ID_BLOCO_2_CH	
- De 1 a 13
#### ID_BLOCO_3_CH	
- De 1 a 13
#### NU_BLOCO_1_ABERTA_CH	
- De 1 a 6
#### NU_BLOCO_2_ABERTA_CH
- De 1 a 6
#### ID_CADERNO_CN	
- Prova Regular (Cadernos 1 a 14)
- Macrotipo 18 (Caderno 15)
#### ID_BLOCO_1_CN	
- De 1 a 7
#### ID_BLOCO_2_CN	
- De 1 a 7
#### ID_BLOCO_3_CN	
- De 1 a 7
#### NU_BLOCO_1_ABERTA_CN	
- De 1 a 3
#### NU_BLOCO_2_ABERTA_CN
- De 1 a 3
#### TX_RESP_BLOCO_1_LP	
- A, B, C, D, . (branco), * (nulo)
#### TX_RESP_BLOCO_2_LP	
- A, B, C, D, . (branco), * (nulo)
#### TX_RESP_BLOCO_1_MT	
- A, B, C, D, . (branco), * (nulo)
#### TX_RESP_BLOCO_2_MT	
- A, B, C, D, . (branco), * (nulo)
#### TX_RESP_BLOCO_1_CH
- A, B, C, D, . (branco), * (nulo)
#### TX_RESP_BLOCO_2_CH	
- A, B, C, D, . (branco), * (nulo)
#### TX_RESP_BLOCO_3_CH	
- A, B, C, D, . (branco), * (nulo)
#### CO_CONCEITO_Q1_CH	
- 0: Nenhum crédito
- 1: Crédito parcial
- 2: Crédito total
- 7: Erros de impressão ou digitalização
- .: Branco
#### CO_CONCEITO_Q2_CH	
- 0: Nenhum crédito
- 1: Crédito parcial
- 2: Crédito total
- 7: Erros de impressão ou digitalização
- .: Branco
#### TX_RESP_BLOCO1_CN	
- A, B, C, D, . (branco), * (nulo)
#### TX_RESP_BLOCO2_CN	
- A, B, C, D, . (branco), * (nulo)
#### TX_RESP_BLOCO3_CN	
- A, B, C, D, . (branco), * (nulo)
#### CO_CONCEITO_Q1_CN	
- 0: Nenhum crédito
- 1: Crédito parcial
- 2: Crédito total
- 7: Erros de impressão ou digitalização
- .: Branco
#### CO_CONCEITO_Q2_CN	
- 0: Nenhum crédito
- 1: Crédito parcial
- 2: Crédito total
- 7: Erros de impressão ou digitalização
- .: Branco
#### IN_PROFICIENCIA_LP	
- 0: Não
- 1: Sim
#### IN_PROFICIENCIA_MT	
- 0: Não
- 1: Sim
#### IN_PROFICIENCIA_CH
- 0: Não
- 1: Sim
#### IN_PROFICIENCIA_CN	
- 0: Não
- 1: Sim
#### IN_AMOSTRA
- 0: Não
- 1: Sim
#### ESTRATO	
- Os estratos são compostos por características da participação da escola na avaliação e representam agrupamentos para os quais a avaliação fornece resultados confiáveis para Língua Portuguesa e Matemática. 
#### ESTRATO_CIENCIAS
- Os estratos são compostos por características da participação da escola na avaliação e representam agrupamentos para os quais a avaliação fornece resultados confiáveis para Ciências Humanas e Ciências da Natureza. 
#### PESO_ALUNO_LP
- Valor com 7 casas decimais 
#### PROFICIENCIA_LP	
- Valor com 7 casas decimais 
#### ERRO_PADRAO_LP
- Valor com 7 casas decimais 
#### PROFICIENCIA_LP_SAEB	
- Valor com 7 casas decimais 
#### ERRO_PADRAO_LP_SAEB	
- Valor com 7 casas decimais 
#### PESO_ALUNO_MT
- Valor com 7 casas decimais 
#### PROFICIENCIA_MT	
- Valor com 7 casas decimais 
#### ERRO_PADRAO_MT	
- Valor com 7 casas decimais 
#### PROFICIENCIA_MT_SAEB	
- Valor com 7 casas decimais 
#### ERRO_PADRAO_MT_SAEB
- Valor com 7 casas decimais 
#### PESO_ALUNO_CH	
- Valor com 7 casas decimais 
#### PROFICIENCIA_CH
- Valor com 7 casas decimais 
#### ERRO_PADRAO_CH	
- Valor com 7 casas decimais 
#### PROFICIENCIA_CH_SAEB	
- Valor com 7 casas decimais 
#### ERRO_PADRAO_CH_SAEB	
- Valor com 7 casas decimais 
#### PESO_ALUNO_CN
- Valor com 7 casas decimais 
#### PROFICIENCIA_CN
- Valor com 7 casas decimais 
#### ERRO_PADRAO_CN
- Valor com 7 casas decimais 
#### PROFICIENCIA_CN_SAEB
- Valor com 7 casas decimais 
#### ERRO_PADRAO_CN_SAEB
- Valor com 7 casas decimais
#### IN_PREENCHIMENTO_QUESTIONARIO
- 0: Não preenchido
- 1: Preenchido parcial ou totalmente
#### TX_RESP_Q001	- Qual língua você fala com mais frequência em sua casa?
- A: Português
- B: Espanhol
- C: Outra língua
#### TX_RESP_Q002 - Qual é a sua cor ou raça?
- A: Branca
- B: Preta
- C: Parda
- D: Amarela
- E: Indígena
- F: Não quero declarar
#### TX_RESP_Q003a - Normalmente, quem mora na sua casa? - Mãe (mães ou madrasta)
- A: Não
- B: Sim
#### TX_RESP_Q003b - Normalmente, quem mora na sua casa? - Pai (pais ou padrasto)
- A: Não
- B: Sim
#### TX_RESP_Q003c -  Normalmente, quem mora na sua casa? - Irmão(s) ou irmã(s)
- A: Não
- B: Sim
#### TX_RESP_Q003d - Normalmente, quem mora na sua casa? - Avô ou avó
- A: Não
- B: Sim
#### TX_RESP_Q003e -  Normalmente, quem mora na sua casa? - Outros (tios, primos, etc.)
- A: Não
- B: Sim
#### TX_RESP_Q004 - Qual é a maior escolaridade da sua mãe (ou mulher responsável por você)?
- A: Não completou o 5° ano do Ensino Fundamental
- B: Ensino Fundamental, até o 5° ano
- C: Ensino Fundamental completo
- D: Ensino Médio completo
- E: Ensino Superior completo (faculdade ou graduação)
- F: Não sei
#### TX_RESP_Q005 - Qual é a maior escolaridade do seu pai (ou homem responsável por você)?
- A: Não completou o 5° ano do Ensino Fundamental
- B: Ensino Fundamental, até o 5° ano
- C: Ensino Fundamental completo
- D: Ensino Médio completo
- E: Ensino Superior completo (faculdade ou graduação)
- F: Não sei
#### TX_RESP_Q006a - Com que frequência seus pais ou responsáveis costumam: - Conversar com você sobre o que acontece na escola
- A: Nunca ou quase nunca
- B: De vez em quando
- C: Sempre ou quase sempre
#### TX_RESP_Q006b -  Com que frequência seus pais ou responsáveis costumam: - Incentivar você a estudar
- A: Nunca ou quase nunca
- B: De vez em quando
- C: Sempre ou quase sempre
#### TX_RESP_Q006c -  Com que frequência seus pais ou responsáveis costumam: - Incentivar você a fazer a tarefa de casa
- A: Nunca ou quase nunca
- B: De vez em quando
- C: Sempre ou quase sempre
#### TX_RESP_Q006d -  Com que frequência seus pais ou responsáveis costumam: - Incentivar você a comparecer às aulas
- A: Nunca ou quase nunca
- B: De vez em quando
- C: Sempre ou quase sempre
#### TX_RESP_Q006e -  Com que frequência seus pais ou responsáveis costumam: - Ir às reuniões de pais na escola
- A: Nunca ou quase nunca
- B: De vez em quando
- C: Sempre ou quase sempre
#### TX_RESP_Q007 - Com que frequência sua família paga alguém para auxiliar nos trabalhos domésticos (faxina ou limpeza)?
- A: Nunca ou quase nunca
- B: De vez em quando (uma vez por semana, a cada quinze dias, etc.)
- C: Sempre ou quase sempre (ex.:três ou mais dias por semana)
#### TX_RESP_Q008a - Na região que você mora tem: Rua pavimentada (asfalto ou calçamento)
- A: Não
- B: Sim
#### TX_RESP_Q008b - Na região que você mora tem: Água tratada da rua
- A: Não
- B: Sim
#### TX_RESP_Q008c - Na rua que você mora tem: Iluminação na rua
- A: Não
- B: Sim
#### TX_RESP_Q009a - Dos itens relacionados abaixo, quantos existem na sua casa? - Geladeira
- A: Nenhum
- B: 1
- C: 2
- D: 3 ou mais
#### TX_RESP_Q009b - Dos itens relacionados abaixo, quantos existem na sua casa? - Tablet
- A: Nenhum
- B: 1
- C: 2
- D: 3 ou mais
#### TX_RESP_Q009c - Dos itens relacionados abaixo, quantos existem na sua casa? - Computador (ou notebook)
- A: Nenhum
- B: 1
- C: 2
- D: 3 ou mais
#### TX_RESP_Q009d - Dos itens relacionados abaixo, quantos existem na sua casa? - Quartos para dormir
- A: Nenhum
- B: 1
- C: 2
- D: 3 ou mais
#### TX_RESP_Q009e - Dos itens relacionados abaixo, quantos existem na sua casa? - Televisão
- A: Nenhum
- B: 1
- C: 2
- D: 3 ou mais
#### TX_RESP_Q009f - Dos itens relacionados abaixo, quantos existem na sua casa? - Banheiro
- A: Nenhum
- B: 1
- C: 2
- D: 3 ou mais
#### TX_RESP_Q009g - Dos itens relacionados abaixo, quantos existem na sua casa? - Carro
- A: Nenhum
- B: 1
- C: 2
- D: 3 ou mais
#### TX_RESP_Q010a - Na sua casa tem: - Tv a cabo (Ex.:Net, Sky, etc.)
- A: Não
- B: Sim
#### TX_RESP_Q010b - Na sua casa tem: - Rede Wi-Fi
- A: Não
- B: Sim
#### TX_RESP_Q010c - Na sua casa tem: - Um quarto só seu
- A: Não
- B: Sim
#### TX_RESP_Q010d - Na sua casa tem: - Mesa para estudar (ou escrivaninha)
- A: Não
- B: Sim
#### TX_RESP_Q010e - Na sua casa tem: - Garagem
- A: Não
- B: Sim
#### TX_RESP_Q010f - Na sua casa tem: - Forno de microondas
- A: Não
- B: Sim
#### TX_RESP_Q010g - Na sua casa tem: - Aspirador de pó
- A: Não
- B: Sim
#### TX_RESP_Q010h - Na sua casa tem: - Máquina de lavar roupa 
- A: Não
- B: Sim
#### TX_RESP_Q010i - Na sua casa tem: - Freezer (independente ou segunda porta da geladeira)
- A: Não
- B: Sim
#### TX_RESP_Q011 - Quanto tempo você demora para chegar à sua escola?
- A: Menos de 30 minutos
- B: Entre 30 minutos e uma hora
- C: Mais de uma hora
#### TX_RESP_Q012 - Considerando a maior distãncia percorrida, normalmente de que forma você chega à sua escola?
- A: À pé
- B: De De ônibus urbano 
- C: De transporte escolar
- D: De barco
- E: De bicicleta
- F: De carro
- G: Outros meios de transporte
#### TX_RESP_Q013 - Com que idade você entrou na escola?
- A: 3 anos ou menos
- B: 4 ou 5 anos 
- C: 6 ou 7 anos
- D: 8 anos ou mais
#### TX_RESP_Q014 - A partir do primeiro ano do ensino fundamental, em que tipo de escola você estudou?
- A: Somente em escola pública
- B: Somente em escola particular
- C: Em escola pública e em escola particular
#### TX_RESP_Q015 - Você já foi reprovado(a)?
- A: Não
- B: Sim, uma vez
- C: Sim, duas vezes ou mais
#### TX_RESP_Q016 - Alguma vez você abandonou a escola deixando de frequentá-la até o final do ano escolar?
- A: Nunca
- B: Sim, uma vez
- C: Sim, duas vezes ou mais
#### TX_RESP_Q017a - Fora da escola em dias de aula, quanto tempo você usa para: - Lazer (TV, internet, jogar bola, música, etc.)
- A: Não uso meu tempo pra isso
- B: Menos de 1 hora
- C: Entre 1 e 2 horas
- D: Mais de 2 horas
#### TX_RESP_Q017b - Fora da escola em dias de aula, quanto tempo você usa para: - Fazer cursos
- A: Não uso meu tempo pra isso
- B: Menos de 1 hora
- C: Entre 1 e 2 horas
- D: Mais de 2 horas
#### TX_RESP_Q017c - Fora da escola em dias de aula, quanto tempo você usa para: - Fazer trabalhos domésticos (lavar louça, limpar quintal, cuidar dos irmãos)
- *: Nulo
- .: Branco
- A: Não uso meu tempo pra isso
- B: Menos de 1 hora
- C: Entre 1 e 2 horas
- D: Mais de 2 horas
#### TX_RESP_Q017d - Fora da escola em dias de aula, quanto tempo você usa para: - Estudar (lição de casa, trabalhos escolares, etc.)
- A: Não uso meu tempo pra isso
- B: Menos de 1 hora
- C: Entre 1 e 2 horas
- D: Mais de 2 horas
#### TX_RESP_Q017e - Fora da escola em dias de aula, quanto tempo você usa para: - Trabalhar fora de casa (recebendo ou não um salário)
- A: Não uso meu tempo pra isso
- B: Menos de 1 hora
- C: Entre 1 e 2 horas
- D: Mais de 2 horas
#### TX_RESP_Q018a - Com que frequência você costuma: - Ler notícias (jornais, revistas, internet, etc.) 
- A: Nunca ou quase nunca
- B: De vez em quando
- C: Sempre ou quase sempre
#### TX_RESP_Q018b - Com que frequência você costuma: - Ler livros que não sejam das matérias escolares
- A: Nunca ou quase nunca
- B: De vez em quando
- C: Sempre ou quase sempre
#### TX_RESP_Q018c - Com que frequência você costuma: - Ler histórias em quadrinhos (mangás, gibis, etc.)
- A: Nunca ou quase nunca
- B: De vez em quando
- C: Sempre ou quase sempre
#### TX_RESP_Q019 - Quando terminar o Ensino Fundamental você pretende
- A: Somente continuar estudando
- B: Somente trabalhar
- C: Continuar estudando e trabalhar
- D: Ainda não sei